In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


### View The Data
I will start by seeing what we got from our data collection. How many rows have location data and what does this data look like.

In [2]:
mydata = "tweets1018913.csv"
df = pd.read_csv(mydata)
df.head(5)

Unnamed: 0                 t_id         s_name  \
0         0.0  1052370940128968705  brandycooklyn   
1         1.0  1052370675632091136      Louroulai   
2         2.0  1052370144180211719        _WayneJ   
3         3.0  1052369947236724736   Imposter_Cat   
4         4.0  1052369944506068992     isaidimhip   

                                              t_text             u_location  \
0  THINGS WRITERS LOVE:\n\n- Panel discussions.\n...   Melbourne, Australia   
1  RT @nelsonlflores: Tips for avoiding imposter ...                    NaN   
2  RT @nelsonlflores: Tips for avoiding imposter ...                     GA   
3  RT @isaidimhip: Anyway........anybody else fee...  impostersyndrome.life   
4  Anyway........anybody else feeling that impost...                    NaN   

                                           image_url               t_date  
0  http://pbs.twimg.com/profile_images/1042972658...  2018-10-17 01:29:18  
1  http://pbs.twimg.com/profile_images/2379147043...  2018-10-17 01:28:15  
2  http://pbs.twimg.com/profile_images/1050602456...  2018-10-17 01:26:09  
3  http://pbs.twimg.com/profile_images/1047849327...  2018-10-17 01:25:22  
4  http://pbs.twimg.com/profile_images/1021102005...  2018-10-17 01:25:21

In [3]:
df.dtypes

Unnamed: 0    float64
t_id           object
s_name         object
t_text         object
u_location     object
image_url      object
t_date         object
dtype: object

In [4]:
df.shape

(26819, 7)

In [5]:
df.apply(lambda x: sum(x.isnull()),axis=0)

Unnamed: 0     286
t_id             0
s_name           0
t_text           0
u_location    4646
image_url        0
t_date           0
dtype: int64

In [6]:

df['u_location'].value_counts(ascending=False).head(20)

impostersyndrome.life    4552
London, England           387
u_location                286
London                    256
Los Angeles, CA           220
United States             178
Chicago, IL               166
Washington, DC            151
Boston, MA                141
New York, NY              140
United Kingdom            138
Houston, TX               124
California, USA           117
UK                        104
Seattle, WA               100
Atlanta, GA                99
Brooklyn, NY               88
San Francisco, CA          85
Philadelphia, PA           83
Portland, OR               81
Name: u_location, dtype: int64

In [7]:
df['u_location'].value_counts(ascending=False).tail(20)

Lahti, Suomi                      1
DC/Boston/New Haven               1
Los                               1
A noisy MDF                       1
Soul searching✨                   1
Beantown                          1
dnd                               1
ptx/frisco/Hunstvegas             1
$C                                1
NorCal                            1
Staten Island, NY.                1
Sunderland UK                     1
Old Dominion University ‘21       1
661-415                           1
Sheffield                         1
IG: julia_victoriaa               1
Retweets are not Endorsements     1
Rockville, MD                     1
Disbelief, USA                    1
Bethlehem, PA                     1
Name: u_location, dtype: int64

## What are we looking for?
Ok, some places are real and some are not. Looking at the data I am going to stick with locations in the USA and locations that have been picked via the drop-down menu on twitter. Here is what they look like:
#### City, State:
Maplewood, NJ
#### State, County:
New Jersey, USA
#### Country:
United States

1. Entries will have a single comma or be "United States"
2. Full city name with a state abbriviation
3. Full state name with a country abbriviation
4. Full country name

#### What we want, is to convert it to one of the following:

Maplewood, New Jersey, USA

New Jersey, USA

USA

## Cleaning
1. Remove empty location rows
2. Convert NJ to New Jersey, USA
3. Convert United States to USA
4. Keep only fileds that have USA in them
5. Removing retweets

In [8]:
# mark if a retweet and then drop
df['get_rt'] = df['t_text'].str.startswith('RT')
df.drop(df[df['get_rt'] == True].index, inplace=True)
df.shape

(7795, 8)

That was a lot of retweets the number of original tweets is 7795. Next we drop all empty locations and take a look at our data again.

In [9]:
df.dropna(subset=['u_location'], how='all', inplace = True)

In [10]:
df.apply(lambda x: sum(x.isnull()),axis=0)

Unnamed: 0    286
t_id            0
s_name          0
t_text          0
u_location      0
image_url       0
t_date          0
get_rt          0
dtype: int64

In [11]:
df.shape

(6561, 8)

We started with a little over 26k tweets and are down to around 7k. Now we will convert NJ to New Jersey, USA and United States to USA. I do this using a dictionary for all of the states.

In [12]:
us_state_abbrev = {
    'AL': 'Alabama, USA',
    'AK': 'Alaska, USA',
    'AZ': 'Arizona, USA',
    'AR': 'Arkansas, USA',
    'CA': 'California, USA',
    'CO': 'Colorado, USA',
    'CT': 'Connecticut, USA',
    'DE': 'Delaware, USA',
    'FL': 'Florida, USA',
    'GA': 'Georgia, USA',
    'HI': 'Hawaii, USA',
    'ID': 'Idaho, USA',
    'IL': 'Illinois, USA',
    'IN': 'Indiana, USA',
    'IA': 'Iowa, USA',
    'KS': 'Kansas, USA',
    'KY': 'Kentucky, USA',
    'LA': 'Louisiana, USA',
    'ME': 'Maine, USA',
    'MD': 'Maryland, USA',
    'MA': 'Massachusetts, USA',
    'MI': 'Michigan, USA',
    'MN': 'Minnesota, USA',
    'MS': 'Mississippi, USA',
    'MO': 'Missouri, USA',
    'MT': 'Montana, USA',
    'NE': 'Nebraska, USA',
    'NV': 'Nevada, USA',
    'NH': 'New Hampshire, USA',
    'NJ': 'New Jersey, USA',
    'NM': 'New Mexico, USA',
    'NY': 'New York, USA',
    'NC': 'North Carolina, USA',
    'ND': 'North Dakota, USA',
    'OH': 'Ohio, USA',
    'OK': 'Oklahoma, USA',
    'OR': 'Oregon, USA',
    'PA': 'Pennsylvania, USA',
    'RI': 'Rhode Island, USA',
    'SC': 'South Carolina, USA',
    'SD': 'South Dakota, USA',
    'TN': 'Tennessee, USA',
    'TX': 'Texas, USA',
    'UT': 'Utah, USA',
    'VT': 'Vermont, USA',
    'VA': 'Virginia, USA',
    'WA': 'Washington, USA',
    'WV': 'West Virginia, USA',
    'WI': 'Wisconsin, USA',
    'WY': 'Wyoming, USA',
    'United States': 'USA'
}

In [13]:
# Convert 2 letter abbrv. to State name, USA
df['u_location'] = df['u_location'].replace(us_state_abbrev, regex=True)

In [15]:
# mark all with the count of USA and grab those with only one.
usa_counts = df['u_location'].str.count('USA')
# usa_counts
df['with_usa'] = np.where((usa_counts == 1),df.u_location, "")

In [16]:
df['with_usa'].value_counts(ascending=False).tail(10)

Bayside, New York, USA                        1
NOLouisiana, USA✈️DC                          1
Knoxville, Tennessee, USA                     1
Oregon, USA                                   1
My Mind, Florida, USA                         1
Columbus, Indiana, USA                        1
Kansas, USA                                   1
Windsor Hills, California, USA                1
Beaufort South Carolina, USA | Oxford U.K.    1
Davenport, Florida, USA                       1
Name: with_usa, dtype: int64

In [17]:
df['with_usa'].value_counts(ascending=False).head(20)

                                  4346
Los Angeles, California, USA       111
New York, New York, USA             81
Chicago, Illinois, USA              70
USA                                 66
California, USA                     63
Oak Park, Illinois, USA             61
Seattle, Washington, USA            61
San Francisco, California, USA      51
Illinois, USA                       46
Brooklyn, New York, USA             42
Louisville, Kentucky, USA           41
New York, USA                       38
Boston, Massachusetts, USA          34
Davis, California, USA              34
District of Columbia, USA           33
Austin, Texas, USA                  32
Portland, Oregon, USA               31
Oakland, California, USA            31
Houston, Texas, USA                 31
Name: with_usa, dtype: int64

#### Evaluate
This is much better but we are still getting some krust, I am going to see if I can clean this up a bit more by removing any entry that has a special char. or does not end in USA

In [18]:
# Must end with USA
end_usa = df['u_location'].str.endswith('USA')
df['with_usa'] = np.where((end_usa == True),df.with_usa, "")
#df['with_usa'].value_counts(ascending=False).tail(120)

In [19]:
# No junk in the location field
df['with_usa'] = df['with_usa'].str.replace('[^a-zA-Z\s,]', '')


In [20]:
# make sure all fields are populated
df.dropna(inplace = True)

In [25]:
df['usa_count'] = df['with_usa'].str.count('USA')
df.drop(df[df['usa_count'] == 0].index, inplace=True)


In [26]:
# we can only have up to two commas, drop rows with more
comma_counts = df['with_usa'].str.count(',')
df['with_comma'] = np.where((comma_counts <= 2),df.with_usa, "")
df.shape


(2134, 11)

This takes us down to 2134 tweets that have pretty clean location information.

#### Split into three different rows
I split the location data into three rows with a split in reverse order. This orders the data with a left justify. I then strip each row of extra spaces and drop the column we do not need. Finally I write out the new csv file.

In [27]:
# split into three fields
df[['Country', 'State', 'City']] = df['with_comma'].apply(lambda x:pd.Series(x.split(",")[::-1]))

In [28]:
# strip extra spaces
df['Country'] = df['Country'].str.strip()
df['State'] = df['State'].str.strip()
df['City'] = df['City'].str.strip()


In [29]:
# drop unused columns
df.drop(['with_usa', 'usa_count', 'with_comma', 'get_rt'], axis=1)


Unnamed: 0                 t_id           s_name  \
7             7.0  1052369469580820481       hayleydenb   
16           16.0  1052367305726865408            molls   
79           79.0  1052357068118740992    Hentai4Dinner   
81           81.0  1052356563690840065        emilymias   
91           91.0  1052354795632250881    DopeBlckQueen   
102           1.0  1052352901648125957        kanavjain   
108           7.0  1052352462844133377            gmksp   
118          17.0  1052349532871188480        jimtoepel   
151          50.0  1052345893200904197       _markryann   
153          52.0  1052345782676975616      IamAleshiaP   
166          65.0  1052341667456606209         mznewman   
172          71.0  1052339756267229186    davidbrunelle   
175          74.0  1052338650355261440        drewgrham   
188          87.0  1052335423878569985     ScottGoyette   
196          95.0  1052333909499944960    nelsonlflores   
211           9.0  1052331246683115521    heybluewrites   
217          15.0  1052330227874062337         siddante   
224          22.0  1052329796124991488   AdamDLindstrom   
248          46.0  1052325511068758016       SawakoSono   
267          65.0  1052321750023692289        tinaatang   
272          70.0  1052321336096374792        anna_boeg   
294          92.0  1052316812153520128    musingvirtual   
295          93.0  1052316404752572429           xnaunx   
319          16.0  1052313748352905216     alexhareland   
337          34.0  1052310999729139718  possiblehorizon   
355          52.0  1052308459364978689    Alcrowldabest   
361          58.0  1052308186366009344         RentPoet   
364          61.0  1052307760614785024           chep2m   
366          63.0  1052307684983222272   AmLitCritTwitt   
382          79.0  1052305194883502080           chep2m   
...           ...                  ...              ...   
26537        20.0  1052709673978273792     OliviakUmana   
26542        25.0  1052708514500673536  constancekassor   
26544        27.0  1052708138330324993      JustinFlood   
26551        34.0  1052705788458913792  AlexaEisenbarth   
26577        60.0  1052698267568562176            urcym   
26579        62.0  1052698153877684224           w9jds_   
26584        67.0  1052697264966303744   adrianafitnall   
26587        70.0  1052696694935224320      MsJohnson16   
26589        72.0  1052695582442024964         TDHRehab   
26593        76.0  1052693895585419264    jamesingramii   
26601        84.0  1052692474898968576        dejbrown_   
26614        97.0  1052689723494420480       GrubFellow   
26615        98.0  1052689444472573952     serenalarkin   
26616        99.0  1052689272900182022    katieboundary   
26628        10.0  1052687067925307392           j_lnch   
26675        57.0  1052671040755777538    AnnieRoonie13   
26697        79.0  1052667871191490560           defjoy   
26698        80.0  1052667869073412096      bdclendenin   
26704        86.0  1052667629771608064       biaMedious   
26707        89.0  1052667045060460548      HRay_Of_Sun   
26735        16.0  1052659576133349377       typedoutco   
26737        18.0  1052659091745562625  neener_baneener   
26741        22.0  1052658233972801542     TeganTrovato   
26746        27.0  1052657044724039687         Kcap1313   
26748        29.0  1052656566313336833          bcjordo   
26758        39.0  1052655395569889280  nickolysalmeida   
26769        50.0  1052653395927023617       ladybotfly   
26782        63.0  1052648501144891393      RNickGorton   
26788        69.0  1052647590091780098          mhstein   
26809        90.0  1052643019856330752          JBRylah   

                                                  t_text  \
7      True story. I had a dream last night that I wa...   
16     @katieferg @kimberlylhudson If you're being st...   
79     Does anybody else know what imposter syndrome ...   
81     I had the privilege today to talk to  @MassArt...   
91     this transition to graduate school has be

In [30]:
df.shape

(2134, 14)

In [31]:
# write out my file of clean tweets
df.to_csv('tweets_clean_1018.csv')